src/library_pthread.js

```diff
@@ -490,6 +490,34 @@ var LibraryPThread = {
     }
   },
 
+  $registerTlsInit: function(tlsInitFunc, moduleExports, metadata) {
+#if DYLINK_DEBUG
+    out("registerTlsInit: " + tlsInitFunc);
+#endif
+#if RELOCATABLE
+    // In relocatable builds, we use the result of calling tlsInitFunc
+    // (`emscripten_tls_init`) to relocate the TLS exports of the module
+    // according to this new __tls_base.
+    function tlsInitWrapper() {
+      var __tls_base = tlsInitFunc();
+#if DYLINK_DEBUG
+      err('tlsInit -> ' + __tls_base);
+#endif
+      for (var sym in metadata.tlsExports) {
+        metadata.tlsExports[sym] = moduleExports[sym];
+      }
+      relocateExports(metadata.tlsExports, __tls_base, true);
+    }
+
+    PThread.tlsInitFunctions.push(tlsInitWrapper);
+    if (runtimeInitialized) {
+      tlsInitWrapper();
+    }
```

Comment:
do we need to both push this function and call it now, if the runtime is initialized? If it's *not* initialized, when would it get called - just at the time the pushed function is?

Reply:
Yes, this is intended.  `PThread.tlsInitFunctions` is for future threads, and the direct call to `tlsInitWrapper` is for the currently running thread.

The previous code did this too, just in different place (library_dylink.js):

```
-        PThread.tlsInitFunctions.push(initTLS);
-        if (runtimeInitialized) {
-          initTLS();
-        }
```

Reply:
If the runtime is not yet initialized then that means the main thread has not yet called `threadInit` and in that case all the `PThread.tlsInitFunctions` will be called when it eventually does:

https://github.com/emscripten-core/emscripten/blob/8770a96c2a7a612a0e37e00370a1b707766c50e8/src/library_pthread.js#L224-L236

Reply:
In this case the check for `runtimeInitialized` is like checking for: "has current/main thread processed PThread.tlsInitFunctions already".   I'll add a comment as such.